cmd/pbstatuser/main.go

```diff
@@ -20,25 +30,173 @@ import (
 	"github.com/rs/zerolog/log"
 )
 
+// This needs to be done in 4 goroutines.
+// 1) Reads messages, calls sources to get auth type and ARN and then enqueue the work
+// in one of three channels (AWS, Azure, GCP)
+// 2) Reads from AWS channel, performs availability check, has configurable throttling.
+// 3) Reads from Azure channel, performs availability check, has configurable throttling.
+// 4) Reads from GCP channel, performs availability check, has configurable throttling.
+
+// Beware there is no database connection, no job queue, most of what api/worker processes have
+// is not available because it should not be needed. The worker will simply use SDKs to check
+// statuses and send out results via kafka.Send function to Sources.
+// Also make sure that all goroutines are closed correctly on context cancel. A WaitGroup
+// would be a really good here: https://gist.github.com/fracasula/b579d52daf15426e58aa133d0340ccb0
+
+const ChannelBuffer = 32
+
+type SourceInfo struct {
+	Authentication *clients.Authentication
+
+	SourceID string `json:"resource_id"`
+}
+
+var (
+	chHeader     = make(chan identity.XRHID, ChannelBuffer)
+	chAws        = make(chan SourceInfo, ChannelBuffer)
+	chAzure      = make(chan SourceInfo, ChannelBuffer)
+	chGcp        = make(chan SourceInfo, ChannelBuffer)
+	chSend       = make(chan kafka.SourceResult, ChannelBuffer)
+	receiverWG   = sync.WaitGroup{}
+	processingWG = sync.WaitGroup{}
+	senderWG     = sync.WaitGroup{}
+)
+
 func processMessage(ctx context.Context, message *kafka.GenericMessage) {
 	logger := ctxval.Logger(ctx)
 	// This is just a test to see if Kafka is configured properly in ephemeral
 	logger.Warn().Msg("RECEIVED KAFKA MESSAGE")
 
-	// TODO: implement the source checking for AWS, Azure, GCP
+	// Get source id
+	asm, err := kafka.NewAvailabilityStatusMessage(message)
+	if err != nil {
+		logger.Warn().Msgf("could not get availability status message %s", err)
+		return
+	}
+
+	sourceId := asm.SourceID
+
+	// Get sources client
+	sourcesClient, err := clients.GetSourcesClient(ctx)
+	if err != nil {
+		logger.Warn().Msgf("could not get sources client %s", err)
+		return
+	}
+
+	// Fetch authentication from Sources
+	authentication, err := sourcesClient.GetAuthentication(ctx, sourceId)
+	if err != nil {
+		if errors.Is(err, clients.NotFoundErr) {
+			logger.Warn().Msgf("not found error: %s", err)
+			return
+		}
+		logger.Warn().Msgf("could not get authentication: %s", err)
+		return
+	}
+
+	s := SourceInfo{
+		Authentication: authentication,
+		SourceID:       sourceId,
+	}
+
+	// Send the identity to the sendResult channel
+	chHeader <- ctxval.Identity(ctx)
```

Comment:
I fail to see how do we synchronize this channel with the messages from the three goroutines.
From what I'm reading there is no certainty that the messages won't get missaligned and Identity for message1 will get assigned to response for message2 and vice versa.
Why can't we pass the identity in SourceInfo and then to SourceResult? It would greatly simplify the synchronization of that info with the given message.

I might be wrong tho, do we have some guidance on what this auth is supposed to achieve or on how it behaves for sources?

Reply:
1. It is possible to pass the identity through sourceInfo and SourceResult but it will change the clean implementation of the generic message. 
Because in one case we are reading the identity from the context and in the latter, we will read the identity from the sourceResult struct. I think it is a bit repetitive to have two functions with the same logic but doable. 
2. I think and @lzap correct me if I am wrong, we can achieve this synchronization using wait groups, waiting until the consume go routine will end, but I am not sure it is a good idea (I believe we are losing parallelism). 
3. What do you mean by auth? the identity header? if so it is needed to authenticate to the sources service. 
I am not sure if there is another usage of it (I'll check that)

@lzap What do you think?

Reply:
1. It is possible to pass the identity through sourceInfo and SourceResult but it will change the clean implementation of the generic message.
Because in one case we are reading the identity from the context and in the latter, we will read the identity from the sourceResult struct. I think it is a bit repetitive to have two functions with the same logic but doable.

2. I think and @lzap correct me if I am wrong, we can achieve this synchronization using wait groups, waiting until the consume go routine will end, but I am not sure it is a good idea (I believe we are losing parallelism).

3. What do you mean by auth? the identity header? if so it is needed to authenticate to the sources service.
I am not sure if there is another usage of it (I'll check that)

@lzap What do you think?

Reply:
Solved, thanks!